### ISSUE TITLE
Importing workflows with starting HttpRequestEvent does not register URL

### ISSUE THREAD

**Issue by willnationsdev** [OrchardCMS/OrchardCore#8771]
### Describe the bug

If you "Export all Workflows" in a deployment plan, save the recipe json file, go to your deployment target site, and use Packages Import to load up the recipe file, it will successfully port the workflows over. However, I noticed that I would get 404 Not Found errors when attempting to post to my workflows after loading the recipe. I cross-referenced the URL/token between sites in my javascript `$.ajax(...)` call and the URL displayed in the edit page for the HttpRequestEvent activity, but they were identical. It was stored properly in the Document table of the database as well. It was only after regenerating the URL and copy-pasting the new URL into my javascript that the event started triggering properly.

So, something about the process of saving an HttpRequestEvent actually generates a route with the token, and while the recipe step is properly reproducing all serialized information, it is not updating the in-memory information of that route. Haven't yet traced what/how that is happening though.

### To Reproduce
1. Create a workflow with a starting HttpRequestEvent activity.
    - In my case I did POST with antiforgery and had a form that submitted to it, but I suspect a GET request with no antiforgery would work too, for easier testing.
2. Create a deployment plan to export workflows. Run it to save a local file of the recipe.
3. Go to another tenant. Import the Package to run the recipe and it should be recreated perfectly in the database/UI.
4. Test the workflow and notice that it DOES NOT handle the request properly, returning a 404 instead.

### Expected behavior
HTTP-related events that generate routes should preserve said routes to make them immediately available during recipe import.

**Comment by Skrypt** [comment]
I think the token is generated by using the current DataProtection key. So, if you are exporting it to a different tenant that doesn't use the some DataProtection key the url won't work.

**Comment by willnationsdev** [comment]
> I think the token is generated by using the current DataProtection key.

Yeah, I just noticed this and was coming to comment, haha. Hmm, so, I suppose the "fix" would be to add a recipe step that refreshes all URLs for such events automatically.

**Comment by Skrypt** [comment]
Maybe or maybe not. Depends if you are exporting to a tenant that uses the same DataProtection key. So, yeah maybe we should have a step.

**Comment by willnationsdev** [comment]
> Personally, I'm using the WorkflowManager to retrieve these workflow url's

Same. I wrote a ILiquidFilter that lets me pass in the name and get the workflow URL (although, geez, your solution was a lot simpler than mine. *Steals* <_<). I could probably open a pull request to add it at some point.

    var servicePath = "{{ Request.Scheme }}://{{ Request.Host }}{{ "My Request Name" | workflow_url }}";
    
Something like that anyway.

**Comment by Skrypt** [comment]
Yeah just `{{ "Your Workflow Name" | workflow_url }}` and 🚀

**Comment by willnationsdev** [comment]
Just an FYI, I'm working on a recipe step for this and will also refactor my liquid filter / probably build a tag helper. When it's all done, I'll contribute it back upstream. :-D

**Comment by Skrypt** [comment]
Yes it's fine.

**Comment by willnationsdev** [comment]
Hmm, so, I've created a RecipeStep, and done all the Deployment stuff, so I can...

1. Create a DeploymentPlan that includes a step to refresh workflow request tokens.
2. Generate a proper GUI for them.
3. Generate a proper recipe file for them.
4. Run the recipe to update the records in the database (confirmed).

I use `IWorkflowTypeStore` for both the recipe step and the `ILiquidFilter`, but the liquid filter seems to provide stale data after I run the recipe because the page template doesn't get the updated URL, despite the database being fully updated.

Is there something I'm missing? Something I need to do to refresh the data in the store? Seems like a YesSql-specific issue.

**Comment by deanmarcussen** [comment]
Regards this there is a proposal for a `Secrets` module here https://github.com/OrchardCMS/OrchardCore/pull/7891, which probably just needs to be finished, which would make these data protected value transparently managed between instances.

It covers the many related issues where data protected values throw exceptions when moved to different instances.

I think from memory it manages the tokens for the workflow request event in a local `IMemoryCache` for fast lookups.

The workflow token is a data protected tokenisation of the workflow id, and the activity id, so while the url is also stored on the workflow itself, that doesn't matter, because any tokenised version that is valid will allow access to the workflow. So long as the local dataprotection keys can decrypt it, i.e. it was produced by the same key instance.

Anyway have a read of the pr, it may help you figure out what you're trying to do. 

(I would need to see the actual code to understand what your liquid filter is doing)

**Comment by willnationsdev** [comment]
@deanmarcussen The code isn't that complicated.

Here's the RecipeStep which is resetting the URLs (tried to copy logic from the `HttpWorkflowController`):

```cs
public class RefreshWorkflowRequestTokensRecipeStep : IRecipeStepHandler
{
    private readonly IWorkflowTypeStore _workflowTypeStore;
    private readonly ISession _session;
    private readonly ISecurityTokenService _securityTokenService;

    public RefreshWorkflowRequestTokensRecipeStep(IWorkflowTypeStore workflowTypeStore,
                                            ISession session,
                                            ISecurityTokenService securityTokenService)
    {
        _workflowTypeStore = workflowTypeStore;
        _session = session;
        _securityTokenService = securityTokenService;
    }

    public async Task ExecuteAsync(RecipeExecutionContext context)
    {
        if (!String.Equals(context.Name, "RefreshWorkflowRequestTokens", StringComparison.OrdinalIgnoreCase))
        {
            return;
        }

        var wts = (await _workflowTypeStore.GetByStartActivityAsync(nameof(HttpRequestEvent))).ToList();
        wts.ForEach(w =>
        {
            var activity = w.Activities.FirstOrDefault(a => a.IsStart);
            if (activity == null)
            {
                return;
            }
            string url = activity.Properties["Url"].ToString();
            var vars = System.Web.HttpUtility.ParseQueryString(url);
            var token = vars?["token"];
            if (!_securityTokenService.TryDecryptToken<WorkflowPayload>(token, out var payload))
            {
                var tokenLifeSpan = Convert.ToInt32(activity.Properties["TokenLifeSpan"].ToString());
                var newToken = _securityTokenService.CreateToken(new WorkflowPayload(w.WorkflowTypeId, activity.ActivityId), TimeSpan.FromDays(tokenLifeSpan));
                var newUrl = Regex.Replace(url, "token=.*", String.Format("token={0}", newToken));
                activity.Properties["Url"] = newUrl;
                activity.Properties["TokenLifeSpan"] = tokenLifeSpan;
                _workflowTypeStore.SaveAsync(w).Wait();
            }
        });
    }
}
```

And here's the LiquidFilter:

```cs
public class WorkflowUrlLiquidFilter : ILiquidFilter
{
    private static string _taskType = nameof(HttpRequestEvent);
    public ValueTask<FluidValue> ProcessAsync(FluidValue input, FilterArguments arguments, TemplateContext ctx)
    {
        if (!ctx.AmbientValues.TryGetValue("Services", out var item) || !(item is IServiceProvider services))
        {
            return ThrowArgumentException<ValueTask<FluidValue>>("Services missing while invoking 'config_value'");
        }
        var workflowName = input.ToStringValue();
        var names = arguments.Names;

        var workflowTypeStore = services.GetRequiredService<IWorkflowTypeStore>();
        var workflowTypes = workflowTypeStore.GetByStartActivityAsync(_taskType).GetAwaiter().GetResult();
        var workflowType = workflowTypes.FirstOrDefault(w => w.Name == workflowName);
        var activity = workflowType?.Activities.FirstOrDefault(a => a.IsStart);
        var url = activity?.Properties["Url"].ToString();
        return new ValueTask<FluidValue>(FluidValue.Create(url ?? ""));
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static T ThrowArgumentException<T>(string message)
    {
        throw new ArgumentException(message);
    }
}
```

So, from what you're telling me, it sounds like there is some sort of `IMemoryCache` somewhere that I must also refresh? I mean, the process of generating a new security token in `HttpWorkflowController.GenerateId()` doesn't do anything different than what I'm doing, so I'm not sure what I'm missing. Like, I'm building the keys correctly as far as I can tell, but then the way my LiquidFilter is fetching the key doesn't appear to be accurate, as if it were taking it from an out-of-date cache. But both of them are just pulling in the `IWorkflowTypeStore` instance as a service, so I'm not sure what the issue is.

**Comment by deanmarcussen** [comment]
No, there is currently no memory cache. That's a feature of the pr I pointed you at, for information.

It would be to prevent what you are having to do here, which is make a database query everytime you need the url (performance implications)

This regex doesn't look quite right. `var newUrl = Regex.Replace(url, "token=.*", String.Format("token={0}", newToken));`

The use of the `.Wait()` and `.GetAwaiter()` in both methods will cause you problems. Suggest just using `await` 

But I don't see a particular reason you are getting the old url, unless it has not correctly been updated in the database.

Suggest looking in the database to be sure that the url value has actually changed.

I'm not 100% confident that `activity.Properties["Url"] = newUrl` will update the entity correctly. Normally we use `.As()` and `.Put()` extensions, though workflows are different because they are not actually an entity in the db, just inside the workflow type. Odd. Check it in the debugger/ database itself, that it has been persisted correctly, would be my suggestion. 

Also @jtkech noted on my other pr (which I need to get back to) that a http request event does not necesarily need to be the starting event, it can come later from a resumed workflow. But again that shouldn't affect you, if all of yours are always starting events.

**Comment by willnationsdev** [comment]
> Suggest looking in the database to be sure that the url value has actually changed.

I confirmed that the database does indeed update. My local Sqlite instance shows the updated data and if I go to the edit page of the activity, it also displays the updated URL. So the data *is* definitely getting updated within the database correctly.

**What IS strange though** (cause I was sure I had checked it), is that the LiquidFilter *is* putting out the same value that is in the database, but the value is still not counted as valid on POST. So despite the fact that I'm generating the token based on the same parameters the HttpWorkflowController is, it doesn't decrypt properly. I can only think that I must just not be building the token correctly somehow.

> The use of the .Wait() and .GetAwaiter() in both methods will cause you problems. Suggest just using await

Mk. Done (though no difference of course).

> I'm not 100% confident that activity.Properties["Url"] = newUrl will update the entity correctly. Normally we use .As() and .Put() extensions

The As and Put extensions don't appear to work when I am trying to extract a `string` or `JValue`. I get a compiler error about those types not having a parameterless constructor and therefore not being valid generic parameters (which...is weird).

> a http request event does not necesarily need to be the starting event, it can come later from a resumed workflow

Ah! Didn't think about that (and haven't had a chance to read the other PR yet). Will have to take a look at that and also revise my stuff to not make that assumption. Still, yes, unrelated to my current issue.

---

Ahhhh, I found it! I just wasn't plugging in the right value for the token lifespan. HttpWorkflowController slips in a large number if the value is 0 (which my activities' are), so I have to do the same thing when extracting any 0 values from the activity entities in the recipe step. Now the liquid filter's generated URL is working properly.

**Comment by weirdyang** [comment]
Is this the right way to register IRecipeHandlers? 

```
     services.AddScoped<IRecipeStepHandler, RefreshWorkflowRequestTokensRecipeStep>();
```

**Comment by Skrypt** [comment]
@willnationsdev Please do a PR :wink:

**Comment by willnationsdev** [comment]
@Skrypt I thought someone was adding a secrets module that would allow the encryption keys to automatically be migrated across tenants without the need for refreshing the tokens. Is that not the case? I have just been using my deployment step as a stopgap until that feature I merged.

**Comment by Skrypt** [comment]
The Workflow Liquid TagHelper?

**Comment by willnationsdev** [comment]
Ohhhh, right, yes. I can probably do, although, I don't recall ever successfully being able to compile the OC repo. Will have to try and get that working first. Don't wanna submit a blind PR. XD

**Comment by weirdyang** [comment]
Just a heads up, this only updates the token. If you wish to update the full URL, e.g.for cases where you change the app name, you will need to generate it using Url.Action, and the area parameter set to OrchardCore.Workflows.

**Comment by effapp** [comment]
Hi,
Has this problem been solved by now?

**Comment by sebastienros** [comment]
@effapp we fixed something similar recently in main, would you like to try and repro the problem to confirm?
